### PR TITLE
Add Gaussian pRF models

### LINKED ISSUE #1: Implement model base classes

### LINKED ISSUE #2: Implement Gaussian pRF model

### LINKED ISSUE #7: Add mypi type checking
First check whether this works with keras.

### PR DESCRIPTION
**Pull request by maltelueken**
Closes #1 
Closes #2 
Closes #7 

Adds:

- A model base class
- A response model base class
- A generic function for Gaussian pRF responses
- A 2D Gaussian Response Model class
- Several exceptions
- A workflow for testing on all three backends
- A worfklow for type checking

I chose to implement the Gaussian response function in an independent function that works for any number of dimensions (tested for 1, 2, and 3 dimensions). That way the response model classes only become a light wrapper around the function.

In the tests, I made heavy use of inheritance and fixtures so that we don't need to repeat creating objects and tests.

Also, numpy is apparently not a valid backend, but is only used for some operations in `keras.ops`.

### CODE REVIEW

**Review by f-hafner — APPROVED** [review]
This looks great in general, and I like the approach with independent function. I also like the checks and corresponding exceptions, though I think we'll have to document this a bit better in tutorials/docs.

I left a few comments, some for docs, some for my understanding. A bigger comment is regarding the tests of the base class.

In the CI, you're installing the backends a couple of times explicitly. I think we had a discussion about something related, but I would consider having the different backends as optional dependencies in pyproject.toml:
```toml
[project.optional-dependencies]
torch = ["torch"]
jax = ["jax"]
tensorflow = ["tensorflow"]
```
This would create a single source of truth which can be referred to both by users (install package with `pip install our-package[torch]`, and in the CI (add the respective backend to the existing calls with `pip install . [...]`.)


Lastly, I think the dunder methods are missing? I'd probably add them to the specific models?

File: tests/models/test_base.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+    with pytest.raises(ParameterSizeError) as excinfo:
+        raise ParameterSizeError(param_names, param_shapes)
+
+    for param_name, param_shape in zip(param_names, param_shapes, strict=False):
+        assert param_name in str(excinfo.value)
+        assert str(param_shape[0]) in str(excinfo.value)
+
+
+class TestBaseModel:
```
[the remark this answers is not in the document]
Comment (maltelueken) [reply tests/models/test_base.py:35]: Agreed, I originally implemented things this way because pylint complained in VS Code locally, but mypy does indeed not complain when abstract methods return `None` when implementations of child classes return a different type. I now return `None` in the abstract classes and do not test them.

File: src/popylar_prf/models/gaussian.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    Parameters
+    ----------
+    grid : Tensor
+        Stimulus grid for which to make predictions.
+    mu : Tensor
+        Centroid of the population receptive field. The first axis corresponds to the number of voxels. The second axis
+        corresponds to the number of grid dimensions and must match the size of the last `grid` axis.
+    sigma : Tensor
+        Size of the population receptive field. The first axis corresponds to the number of voxels and the size must
+        match the size of the first `mu` axis.
```
[the remark this answers is not in the document]
Comment (maltelueken) [reply src/popylar_prf/models/gaussian.py]: No, but size is used to refer to dimensions.

I added that the size of the second `sigma` dimension must be one.

**Comment by maltelueken** [comment]
Thanks for the review! I addressed the issue with the abstract methods in the base classes. I also added the backends as optional dependencies.

I also changed the term *voxels* to *batches* as the batch dimension (the first axis) does not always have to refer to voxels (e.g., for connective field modeling). We need to definitely clarify this in tutorials along with the exepctions. Perhaps we can do a FAQ section for exceptions.
